### PR TITLE
Int8 sweeps logits lax parens

### PR DESCRIPTION
**Pull request by gobbleturk**
Initial logits PR

### CODE REVIEW

File: MaxText/layers.py [hunk 1]
```diff
@@ -704,8 +705,26 @@ def attend(self, query: Array) -> Array:
       in NLP models.
     """
     dtype = self.attend_dtype if self.attend_dtype is not None else self.dtype
-    return jnp.dot(query, jnp.asarray(self.embedding, dtype).T)
```
Comment (lukaszlew) [inline MaxText/layers.py:707]: I think it is worthwhile to just fork jnp.dot and add _dot_general parameter to it. That will make the review easier and will be easier to go back to jnp.dot once _dot_general param lands in jax

File: MaxText/layers.py [hunk 2]
```diff
@@ -704,8 +705,26 @@ def attend(self, query: Array) -> Array:
...
-
+    if not self.config.int8_training:
+      #return jnp.dot(query, jnp.asarray(self.embedding, dtype).T)
+      return lax.dot_general(query, jnp.asarray(self.embedding, dtype).T, (((query.ndim - 1,),(0,)),((),())))
+    else:
+      aqt_cfg = maxtext_sweeps.sweep1(
+        self.config.fwd_int8_qk,
```
Comment (lukaszlew) [inline MaxText/layers.py:713]: why qk? why not _logits

File: MaxText/layers.py [hunk 3]
```diff
@@ -704,8 +705,26 @@ def attend(self, query: Array) -> Array:
...
+        use_fwd_quant=self.config.aqt_use_fwd_quant,
+      )
+      aqt_dot_general = aqt.make_dot_general(aqt_cfg)
+      aqt_key = self.make_rng('aqt')
+      context = aqt.Context(key=aqt_key, train_step=None)
+      aqt_dot_general = functools.partial(aqt_dot_general, context=context)
+      # !!! AQT requires LHS dtype equals RHS dtype
+      # return aqt_dot_general(query, jnp.asarray(self.embedding).T, (((query.ndim - 1,),(0,)),((),())))
+      # The below fails with jnp.bfloat16
+      return aqt_dot_general(jnp.asarray(query, jnp.float32), jnp.asarray(self.embedding, jnp.float32).T, (((query.ndim - 1,),(0,)),((),())))
```
Comment (lukaszlew) [inline MaxText/layers.py:727]: We need to find common type, if both are f32, we are good, if bothr are bf16 we are good.
If they are bf16+f32 we need to cast both to f32

**Review by lukaszlew — CHANGES_REQUESTED** [review]

File: MaxText/layers.py (hunk 2 above)
Comment (gobbleturk) [reply MaxText/layers.py]: Fixed.

File: MaxText/layers.py (hunk 1 above)
Comment (gobbleturk) [reply MaxText/layers.py:707]: Done.

File: MaxText/layers.py (hunk 3 above)
Comment (gobbleturk) [reply MaxText/layers.py]: Done.

File: MaxText/my_jnp_dot.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+from jax._src.typing import Array, ArrayLike, DuckTypedArray, DTypeLike, Shape
+from typing import (Any, Callable, Optional, TypeVar, Union,
+                    cast as type_cast, overload)
+
+def dot(lhs, rhs, _dot_general, precision = None,
```
Comment (lukaszlew) [inline MaxText/my_jnp_dot.py:5]: _dot_general = lax.dot_general default argument would work well

**Review by lukaszlew — APPROVED** [review]

File: MaxText/my_jnp_dot.py (hunk 4 above)
Comment (gobbleturk) [reply MaxText/my_jnp_dot.py]: Done.
